include/fi_util.h

```diff
@@ -378,5 +386,27 @@ void fi_fabric_insert(struct util_fabric *fabric);
 struct util_fabric *fi_fabric_find(const char *name);
 void fi_fabric_remove(struct util_fabric *fabric);
 
+/*
+ * Layered Providers
+ */
+
+typedef struct fi_info* (*utilx_tr_base_info_t)(struct fi_info *base_info);
+typedef struct fi_info* (*utilx_tr_layer_info_t)(struct fi_info *layer_info);
+
+int utilx_getinfo(uint32_t version, const char *node, const char *service,
+			uint64_t flags, const struct fi_provider *prov,
+			const struct fi_info *prov_info, struct fi_info *hints,
+			utilx_tr_layer_info_t tr_layer_info,
+			utilx_tr_base_info_t tr_base_info,
+			int get_base_info, struct fi_info **info);
+int utilx_parse_name(const char *name, int max_tok, char **tok, int get_prefix);
+int utilx_tr_layer_domain_attr(struct fi_domain_attr *layer_attr, char *prefix,
+		struct fi_domain_attr *base_attr);
+int utilx_tr_layer_fabric_attr(struct fi_fabric_attr *layer_attr,
+		struct fi_fabric_attr *base_attr);
+int utilx_tr_base_domain_attr(struct fi_domain_attr *base_attr, char *prefix,
+		struct fi_domain_attr *layer_attr);
+int utilx_tr_base_fabric_attr(struct fi_fabric_attr *base_attr, char *prefix,
+		struct fi_fabric_attr *layer_attr);
```

Comment:
I would prefer a different prefix for these.  'utilx_' is a little long, but 'utilx_tr_base_' or 'util_tr_layer_' it's too much.  (I don't even know what the 'tr' stands for.)  And the two sets of calls take the same arguments, so it's unclear which call to make.

My vote for internal APIs that do not belong to any specific provider is 'ofi'.  Please look at the naming of the existing attr related calls and use either 'check' or 'alter' in the name, based on the behavior.  It may also be possible to modify the existing calls, rather than introducing new internal APIs.